scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1515 @@
+    k : int
+        The column of the elementary permutation.
+    disc : float
+        Centered discrepancy of the design before permutation.
+
+    Returns
+    -------
+    discrepancy : float
+        Centered discrepancy.
+
+    References
+    ----------
+    .. [1] Jin et al. "An efficient algorithm for constructing optimal design
+       of computer experiments", Journal of Statistical Planning and
+       Inference, 2005.
+
+    """
+    sample = np.asarray(sample)
+    n = sample.shape[0]
+
+    z_ij = sample - 0.5
+
+    # Eq (19)
+    c_i1j = 1. / n ** 2. * np.prod(0.5 * (2. + abs(z_ij[i1, :]) +
+                                          abs(z_ij) -
+                                          abs(z_ij[i1, :] - z_ij)),
+                                   axis=1)
+    c_i2j = 1. / n ** 2. * np.prod(0.5 * (2. + abs(z_ij[i2, :]) +
+                                          abs(z_ij) -
+                                          abs(z_ij[i2, :] - z_ij)),
+                                   axis=1)
+
+    # Eq (20)
+    c_i1i1 = (1. / n ** 2 * np.prod(1 + abs(z_ij[i1, :])) -
+              2. / n * np.prod(1. + 0.5 * abs(z_ij[i1, :]) -
+                               0.5 * z_ij[i1, :] ** 2))
+    c_i2i2 = (1. / n ** 2 * np.prod(1 + abs(z_ij[i2, :])) -
+              2. / n * np.prod(1. + 0.5 * abs(z_ij[i2, :]) -
+                               0.5 * z_ij[i2, :] ** 2))
+
+    # Eq (22), typo in the article in the denominator i2 -> i1
+    num = (2 + abs(z_ij[i2, k]) + abs(z_ij[:, k]) -
+           abs(z_ij[i2, k] - z_ij[:, k]))
+    denum = (2 + abs(z_ij[i1, k]) + abs(z_ij[:, k]) -
+             abs(z_ij[i1, k] - z_ij[:, k]))
+    gamma = num / denum
+
+    # Eq (23)
+    c_p_i1j = gamma * c_i1j
+    # Eq (24)
+    c_p_i2j = c_i2j / gamma
+
+    alpha = (1 + abs(z_ij[i2, k])) / (1 + abs(z_ij[i1, k]))
+    beta = (2 - abs(z_ij[i2, k])) / (2 - abs(z_ij[i1, k]))
+
+    g_i1 = np.prod(1. + abs(z_ij[i1, :]))
+    g_i2 = np.prod(1. + abs(z_ij[i2, :]))
+    h_i1 = np.prod(1. + 0.5 * abs(z_ij[i1, :]) - 0.5 * (z_ij[i1, :] ** 2))
+    h_i2 = np.prod(1. + 0.5 * abs(z_ij[i2, :]) - 0.5 * (z_ij[i2, :] ** 2))
+
+    # Eq (25), typo in the article g is missing
+    c_p_i1i1 = ((g_i1 * alpha) / (n ** 2) - 2. * alpha * beta * h_i1 / n)
+    # Eq (26), typo in the article n ** 2
+    c_p_i2i2 = ((g_i2 / ((n ** 2) * alpha)) - (2. * h_i2 / (n * alpha * beta)))
+
+    # Eq (26)
+    sum_ = c_p_i1j - c_i1j + c_p_i2j - c_i2j
+
+    mask = np.ones(n, dtype=bool)
+    mask[[i1, i2]] = False
+    sum_ = sum(sum_[mask])
+
+    disc_ep = (disc + c_p_i1i1 - c_i1i1 + c_p_i2i2 - c_i2i2 + 2 * sum_)
+
+    return disc_ep
+
+
+def primes_from_2_to(n):
+    """Prime numbers from 2 to *n*.
+
+    Parameters
+    ----------
+    n : int
+        Sup bound with ``n >= 6``.
+
+    Returns
+    -------
+    primes : list(int)
+        Primes in ``2 <= p < n``.
+
+    References
+    ----------
+    .. [1] `StackOverflow <https://stackoverflow.com/questions/2068372>`_.
```

Comment:
@rgommers it looks like the code here is a refactored version of the code in a stackoverflow post from 1/3/2019. 
The license for posts made after 2018-05-02 UTC is [CC BY-SA 4.0](https://creativecommons.org/licenses/by-sa/4.0/), which states that "you must distribute your contributions under the same license as the original." Can we use this?

Reply:
Quick search in scipy's code base shows lot's of comments about code borrowed from SO.

Reply:
We can certainly use it I think. The SO license change was a bit unfortunate. For small code snippets that often doesn't matter (a one-liner isn't copyrightable, the boundary is fuzzy). It may be good to just add explicitly in the Notes section "taken from [1] by <name>, licensed under <license `link`>".

Reply:
Ok I will add a more comprehensive note as you suggest.

Reply:
@rgommers If we think the small snippet isn't copyrightable, then we can use it (attributing it out of courtesy and auditability). But in that case, we shouldn't say that's its incarnation here is licensed under CC-BY-SA. If we do think it's copyrightable, then we need to obey its license, which means that we need to impose the SA conditions on scipy's users, which is incompatible with scipy's general licensing terms.

@tupui I recommend seeking the original author's permission to include this snippet under the terms of scipy's license or find another sieve implementation.

Reply:
@rkern I sent an email to the author from SO (you are in copy).

Reply:
> @rgommers If we think the small snippet isn't copyrightable, then we can use it (attributing it out of courtesy and auditability). But in that case, we shouldn't say that's its incarnation here is licensed under CC-BY-SA. 

This is a good point actually. In the current issue we have with the Matlab code, the underlying principle is published and known. The code is not doing more, hence from what you say we should just remove the reference and use it as we please.

Reply:
In such a case, the best practice that we prefer is to start with the paper and implement from scratch rather than copying/adapting the code. The code may well be copyrightable even if it is a straightforward implementation of pseudocode from a paper.

Reply:
> But in that case, we shouldn't say that's its incarnation here is licensed under CC-BY-SA.

Indeed, I made a mistake there.

Looks like we received the author's permission now, so we are fine.

Reply:
I will do a PR to add a comment here with the consent from the author. Also I will add something in our license section of the doc about not copy-pasting StackOverflow code without consent.